config-linux: add CFS bandwidth burst (#1120)
opencontainers/runtime-spec: config-linux.md

The following parameters can be specified to set up the controller:

* **`shares`** *(uint64, OPTIONAL)* - specifies a relative share of CPU time available to the tasks in a cgroup
* **`quota`** *(int64, OPTIONAL)* - specifies the replenished amount of time in microseconds for which all tasks in a cgroup can run during one period (as defined by **`period`** below)
    A value of `-1` indicates that the cgroup does not have any bandwidth restriction in place (i.e., an unconstrained bandwidth group).
    Any (valid) positive value(s) no smaller than `burst` will enact the specified bandwidth limit.
* **`burst`** *(uint64, OPTIONAL)* - specifies the maximum amount of accumulated time in microseconds for which all tasks in a cgroup can run additionally for burst during one period (as defined by **`period`** below)
    A value of `0` for indicates that the cgroup can not accumulate any unused bandwidth. It makes the traditional bandwidth control behavior unchanged.

>     Any (valid) positive value(s) no larger than `quota` will enact the cap on unused bandwidth accumulation.

Reviewer: Just a question on this line (actually the original linux commit (https://github.com/torvalds/linux/commit/f418371#diff-cc1a82129952a910fdc4292448c2a097a2ba538bebefcf3c06381e45639ae73eR9814), why values larger than quota should return EINVAL? Do you know the reason?

Author: @m-yosefpor Thanks for the question!

In short, a burst value much larger than the quota is breaking the design purpose of bwc (as isolation). The limit put on the burst aims to control the interference to other groups.

FYI, this was discussed in the v4 patch set of this feature: https://lwn.net/ml/linux-kernel/YFG4hEOe65cbCo26@hirez.programming.kicks-ass.net/.

Reviewer: @kailun-qin Thanks for your explanation. Completely makes sense.

The following parameters can be specified to set up the controller:

* **`shares`** *(uint64, OPTIONAL)* - specifies a relative share of CPU time available to the tasks in a cgroup

> * **`quota`** *(int64, OPTIONAL)* - specifies the replenished amount of time in microseconds for which all tasks in a cgroup can run during one period (as defined by **`period`** below)

Reviewer 2: IMO, "the total amount ... during one period" already implies that this is "replenished" and changing this word makes it less clear :see_no_evil:

>     A value of `-1` indicates that the cgroup does not have any bandwidth restriction in place (i.e., an unconstrained bandwidth group).

Reviewer 3: I am not sure about having `-1` here.  On cgroup v2 `quota` and `period` are written together.  Does `-1` mean it has the same value as `period`?

Author: For cgroup v2, I suppose `-1` means `max`, which can be written together w/ `period`.

According to the the `cpu.max` section of [cgroup v2 doc](https://github.com/torvalds/linux/blob/master/Documentation/admin-guide/cgroup-v2.rst#cpu):
```
cpu.max
A read-write two value file which exists on non-root cgroups. The default is "max 100000".

The maximum bandwidth limit. It's in the following format:
$MAX $PERIOD
which indicates that the group may consume upto $MAX in each $PERIOD duration. "max" for $MAX indicates no limit. If only one number is written, $MAX is updated.
````

Author: I think this applies to all negative integers actually (i.e., not only `-1`). Any suggestions? Thanks!

Reviewer 2: I would guess the question here is really why it's important to `burst` support that we include support for `-1` here too -- I'd say this is a breaking change, since the spec doesn't currently say anything about negative numbers, and instead implies that this optional field be left unspecified for no restriction.  Does `cgroupv1` support negative values?

To be clear, I agree that we should clarify this, especially that unspecified needs to be `max` in `cgroupv2`, but I'm not sure how it relates to `burst` except that it's adjacent.

>     A value of `-1` indicates that the cgroup does not have any bandwidth restriction in place (i.e., an unconstrained bandwidth group).

Author: `quota` is defined to be in int64 here, so it's OCI runtime implementation's responsiblity to handle this differently for v1 and v2.

>     Any (valid) positive value(s) no smaller than `burst` will enact the specified bandwidth limit.

Reviewer 2: My understanding so far is that `burst` is either excess time "earned" from past under-usage or borrowed from future periods where we expect/hope we'll be under-utilizing (I'm not actually clear from this PR which it is).  Either way, in this specific case, I think the restriction is really on the value of _`burst`_, not a restriction on the value of `quota` (so doesn't seem helpful to mention here).

> * **`burst`** *(uint64, OPTIONAL)* - specifies the maximum amount of accumulated time in microseconds for which all tasks in a cgroup can run additionally for burst during one period (as defined by **`period`** below)

Reviewer 4: ```suggestion
* **`burst`** *(uint64, OPTIONAL)* - specifies the maximum amount of accumulated time in microseconds for which all tasks in a cgroup can run additionally for burst during one period (as defined by **`period`** below). This feature just be supported in the Linux 5.14 or newer
```

Author: Normally we do not highlight the kernel version requirement in runtime spec.

Reviewer 4: OK, Got it

Reviewer 2: Personally, I think it would be valuable for us to list the relevant kernel version that added each feature (above some global minimum we assume, which I don't think is currently codified anywhere in the spec even though we definitely do assume _some_ minimum), but I agree that it doesn't make sense to do it just for this one instance. :smile:

> * **`burst`** *(uint64, OPTIONAL)* - specifies the maximum amount of accumulated time in microseconds for which all tasks in a cgroup can run additionally for burst during one period (as defined by **`period`** below)

Reviewer 2: > This feature borrows time now against the future underrun ...

Given this, I don't think "accumulated" is the right choice of word here -- this feature is about borrowing from the future, not accumulating from the past:

```suggestion
* **`burst`** *(uint64, OPTIONAL)* - specifies the maximum amount of burst time in microseconds for which all tasks in a cgroup can borrow from future periods during one period (as defined by **`period`** below)
```

Author: `This feature borrows time now against the future underrun ...` is directly from the [kernel doc](https://www.kernel.org/doc/html/latest/scheduler/sched-bwc.html) and it's actually `accumulating from the past` rather than borrowing from future. It intends to say that `borrowing time now to prepare for the future underestimation`.

>     A value of `0` for indicates that the cgroup can not accumulate any unused bandwidth. It makes the traditional bandwidth control behavior unchanged.
>     Any (valid) positive value(s) no larger than `quota` will enact the cap on unused bandwidth accumulation.

Reviewer 2: I think this would be more consistent if we assume that not specifying this optional field is the way to get "no limit", and I think it would also make sense to be more clear about values bigger than `quota` are not just ignored but rather an error, perhaps something like this?

```suggestion
    If specified, this value MUST be no larger than `quota` (runtimes MAY generate an error).
```

**Edit:** coming back to this after https://github.com/opencontainers/runc/pull/3205#discussion_r705484770, I'm guessing that's related? (And likely why the "negative values" bits were added to `quota`?)
